### PR TITLE
feat(credential keyvalue pair): change credential field to type password

### PR DESCRIPTION
**Pull request by Mohammer5**
This will make sure that credential fields of key value pairs are of type password and adds a confirmation field.
I had to copy the `createEqualTo` form validator's code as it doesn't resolve paths, just play keys, while final form can handles paths. I'll create a UI issue for that, we could copy the code I've added in this PR to the UI's validator as well.

### CODE REVIEW

File: src/forms/createEqualTo.js [hunk 1]
```diff
@@ -0,0 +1,52 @@
...
+        { otherField: description || key }
+    )
+
+    return (value, allValues) => {
+        const comparisonValue = getByPath(key, allValues)
+        return isEmpty(value) || value === comparisonValue
+            ? undefined
+            : errorMessage
+    }
+}
```
Comment (HendrikThePendric) [inline src/forms/createEqualTo.js:52]: Could well be that I am missing some nuances here, but I'm not sure you needed to craft these validators by hand. In `@dhis2/forms` we have an `isEmpty` and `createEqualTo` validator too. Did these have some issues? Or does this app need a unique implementation?

If the latter is the case, I think it'd be good to give these validators some distinguishing names (i.e. prevent naming overlap with the validators in @dhis2/forms).

File: src/gateways/GatewayKeyValuePairConfidentialValueFields.js [hunk 2]
```diff
@@ -0,0 +1,62 @@
...
+
+const { Field } = ReactFinalForm
+
+const createEqualToConfidentialValue = index =>
+    createEqualTo(
+        console.log(
+            'createFieldGatewayKeyValuePairValueName(index)',
+            createFieldGatewayKeyValuePairValueName(index)
+        ) || createFieldGatewayKeyValuePairValueName(index),
+        FIELD_GATEWAY_KEY_VALUE_PAIR_VALUE_LABEL
```
Comment (HendrikThePendric) [inline src/gateways/GatewayKeyValuePairConfidentialValueFields.js:28]: Pretty sure we would want to get rid of the console.log here...

**Review by HendrikThePendric — COMMENTED** [review]
I didn't have time to test this locally because an urgent issue popped up, but I did leave some comments.

File: src/forms/createEqualTo.js (hunk 1 above)
Comment (Mohammer5) [reply src/forms/createEqualTo.js:52]: `isEmpty` is an internal implementation of the forms library, it's not exposed. I tried to use createEqualTo, but it doesn't work with paths (e. g. `specialChars[1].value`), which is why I added this one. I kept using the same names, as ideally, we port this to the prop types library in a follow up PR

File: src/gateways/GatewayKeyValuePairConfidentialValueFields.js (hunk 2 above)
Comment (Mohammer5) [reply src/gateways/GatewayKeyValuePairConfidentialValueFields.js:28]: yep...

File: src/forms/createEqualTo.js (hunk 1 above)
Comment (HendrikThePendric) [reply src/forms/createEqualTo.js:52]: 👍  makes perfect sense! TBH, I don't even think I realised we should keep the `paths` notation in mind when I wrote those validators.

File: src/gateways/GatewayKeyValuePairValueField.js [hunk 3]
```diff
@@ -0,0 +1,40 @@
...
+const isStringWithLengthAtLeastOne = composeValidators(string, hasValue)
+
+export const createFieldGatewayKeyValuePairValueName = index =>
+    `parameters[${index}].value`
+
+export const FIELD_GATEWAY_KEY_VALUE_PAIR_VALUE_LABEL = i18n.t('Value')
+
+export const GatewayKeyValuePairValueField = ({ index, isConfidential }) => (
+    <Field
+        type={isConfidential ? 'password' : 'input'}
```
Comment (HendrikThePendric) [inline src/gateways/GatewayKeyValuePairValueField.js:24]: this is causing a props error: should be `password` or `text`
(not input)

**Review by HendrikThePendric — COMMENTED** [review]
OK. I found some time to run the app locally and finish my review...

There are a few things that I thin need fixing. I'm just going to _comment_, and not _request changes_ because I'm off tomorrow and I don't want to block this PR from getting merged once these issues have been addressed:
- Remove the console.log from the validator
- Fix the prop-types error

Apart from that it was also still impossible to fully test this functionality because the webapi is still sending empty values ATM. Having said that, everything looks good and I don't see any reason why this wouldn't work.

👍

File: src/gateways/GatewayKeyValuePairConfidentialValueFields.js (hunk 2 above)
Comment (Mohammer5) [reply src/gateways/GatewayKeyValuePairConfidentialValueFields.js]: removed

File: src/gateways/GatewayKeyValuePairValueField.js (hunk 3 above)
Comment (Mohammer5) [reply src/gateways/GatewayKeyValuePairValueField.js]: changed

**Review by HendrikThePendric — APPROVED** [review]
I couldn't test the full functionality, but once the server starts sending these values I can't see why this wouldn't work the same way as the other fields. What I could test worked well, so I'm approving.
